### PR TITLE
[Embedding] Add GPU apply Ops of Adam, AdamAsync and AdamW for multi-…

### PR DESCRIPTION
**Pull request by lixy9474**
…tier storage of EmbeddingVariable.

### CODE REVIEW

File: tensorflow/core/kernels/training_ali_ops.cc [hunk 1]
```diff
@@ -2173,11 +2463,129 @@ TF_CALL_double(REGISTER_CPU_KERNELS);
...
+                      (void*)&dev_v_ptr, (void*)&grad_base,
+                      (void*)&lr, (void*)&beta1, (void*)&beta2,
+                      (void*)&epsilon, (void*)&embedding_dim,
+                      (void*)&task_size};
+
+      cudaLaunchKernel(
+          (void *)SparseApplyAdamAsyncSparseRmspropGPU<T>,
+          (task_size + block_size - 1) / block_size * embedding_dim,
+          block_size, args, 0, nullptr);
+      cudaDeviceSynchronize();
```
Comment (candyzone) [inline tensorflow/core/kernels/training_ali_ops.cc:2574]: necessary?

File: tensorflow/core/kernels/training_ali_ops.cc [hunk 2]
```diff
@@ -2258,36 +2666,138 @@ class KvSparseApplyAdamAsyncOpGPU : public OpKernel {
...
+        functor::KvSparseApplyAdamAsync<Device, T, Tindex, Tstep>()(
+          device, var, m, v, beta1_power.scalar<T>(), beta2_power.scalar<T>(),
+          indices.vec<Tindex>(), grad.flat_outer_dims<T>(), lr.scalar<T>(),
+          beta1.scalar<T>(), beta2.scalar<T>(), epsilon.scalar<T>(),
+          global_step.scalar<Tstep>(), apply_sparse_rmsprop_, inner_dim,
+          ctx->get_allocator(AllocatorAttributes())));
+      } else {
+        auto indices_vec = indices.vec<Tindex>();
+        auto grad_flat = grad.flat_outer_dims<T>();
+        int64 gs = global_step.scalar<int64>()();
```
Comment (candyzone) [inline tensorflow/core/kernels/training_ali_ops.cc:2688]: int64 - > TStep

File: tensorflow/core/framework/embedding/batch.cu.cc [hunk 3]
```diff
@@ -82,6 +82,124 @@ TF_CALL_double(REGISTER_KERNELS_ALL_INDEX)
...
+    *beta2_power_ptr *= beta2;
+  }
+}
+
+template __global__ void SparseApplyAdamAsyncGPU<float>(
+    float**, float**, float**, float*, float,
+    float, float, float, float*, float*, int, long long int);
+template __global__ void SparseApplyAdamAsyncGPU<double>(
+    double**, double**, double**, double*, double,
+    double, double, double, double*, double*, int, long long int);
```
Comment (candyzone) [inline tensorflow/core/framework/embedding/batch.cu.cc:148]: define macro

File: tensorflow/core/kernels/training_ali_ops.cc [hunk 4]
```diff
@@ -2617,4 +3114,300 @@ TF_CALL_float(REGISTER_CPU_KERNELS);
...
+
+    cudaMemcpy(dev_var_ptr, var_ptr,
+               sizeof(T*) * task_size,
+               cudaMemcpyHostToDevice);
+    cudaMemcpy(dev_m_ptr, m_ptr,
+               sizeof(T*) * task_size,
+               cudaMemcpyHostToDevice);
+    cudaMemcpy(dev_v_ptr, v_ptr,
+               sizeof(T*) * task_size,
+               cudaMemcpyHostToDevice);
```
Comment (candyzone) [inline tensorflow/core/kernels/training_ali_ops.cc:3207]: 这三个memcpy有异步优化空间吗

File: tensorflow/core/kernels/training_ali_ops.cc [hunk 5]
```diff
@@ -1475,6 +1475,296 @@ TF_CALL_float(REGISTER_CPU_KERNELS);
...
+ private:
+  bool use_exclusive_lock_;
+  ThreadCopyIdAllocator* thread_copy_id_alloc_ = nullptr;
+  std::function<T*(T*, Tindex, int64, int64, int64)> get_default_v_fn_;
+};
+
+#define REGISTER_KERNELS(T, Tindices)                                 \
+  REGISTER_KERNEL_BUILDER(Name("KvResourceSparseApplyAdam")             \
+                              .Device(DEVICE_GPU)                     \
+                              .HostMemory("indices")                 \
```
Comment (candyzone) [inline tensorflow/core/kernels/training_ali_ops.cc:1734]: indices place到gpu，否则单HBM不可用

File: tensorflow/core/framework/embedding/batch.cu.cc (hunk 3 above)
Comment (lixy9474) [reply tensorflow/core/framework/embedding/batch.cu.cc]: done

File: tensorflow/core/kernels/training_ali_ops.cc (hunk 2 above)
Comment (lixy9474) [reply tensorflow/core/kernels/training_ali_ops.cc]: done

File: tensorflow/core/kernels/training_ali_ops.cc (hunk 4 above)
Comment (candyzone) [reply tensorflow/core/kernels/training_ali_ops.cc:3207]: optimize later

**Review by candyzone — APPROVED** [review]

**Pull request merged**